doc.go

```diff
@@ -4,6 +4,309 @@
+//
+// It is recommended to use the value set in the Cassandra config for broadcast_address or listen_address,
+// an IP address not a domain name. This is because events from Cassandra will use the configured IP
+// address, which is used to index connected hosts. If the domain name specified resolves to more than 1 IP address
+// then the driver may connect multiple times to the same host, and will not mark the node being down or up from events.
+//
+// Then you can customize more options (see ClusterConfig):
+//
+//  cluster.Keyspace = "example"
+//  cluster.Consistency = gocql.Quorum
+//  cluster.ProtocolVersion = 4
+//
+// The driver tries to automatically detect the protocol version to use if not set, but you might want to set the
+// protocol version explicitly, as it's not defined which version will be used in certain situations (for example
+// during upgrade of the cluster when some of the nodes support different set of protocol versions than other nodes).
+//
+// When ready, create a session from the configuration. Don't forget to Close the session once you are done with it:
+//
+//  session, err := cluster.CreateSession()
+//  if err != nil {
+//  	return err
+//  }
+//  defer session.Close()
+//
+// Authentication
+//
+// CQL protocol uses a SASL-based authentication mechanism and so consists of an exchange of server challenges and
+// client response pairs. The details of the exchanged messages depend on the authenticator used.
+//
+// To use authentication, set ClusterConfig.Authenticator or ClusterConfig.AuthProvider.
+//
+// PasswordAuthenticator is provided to use for username/password authentication:
+//
+//  cluster := gocql.NewCluster("192.168.1.1", "192.168.1.2", "192.168.1.3")
+//  cluster.Authenticator = gocql.PasswordAuthenticator{
+//		Username: "user",
+//		Password: "password"
+//  }
+//  session, err := cluster.CreateSession()
+//  if err != nil {
+//  	return err
+//  }
+//  defer session.Close()
+//
+// Transport layer security
+//
+// It is possible to secure traffic between the client and server with TLS.
+//
+// To use TLS, set the ClusterConfig.SslOpts field. SslOptions embeds *tls.Config so you can set that directly.
+// There are also helpers to load keys/certificates from files.
+//
+// Warning: Due to backward-compatibility reasons, the tls.Config's InsecureSkipVerify is set to
+// !SslOptions.EnableHostVerification, so by default host verification is disabled. Most users using TLS should set
+// SslOptions.EnableHostVerification to true.
+//
+//  cluster := gocql.NewCluster("192.168.1.1", "192.168.1.2", "192.168.1.3")
+//  cluster.SslOpts = &gocql.SslOptions{
+//  	EnableHostVerification: true,
+//  }
+//  session, err := cluster.CreateSession()
+//  if err != nil {
+//  	return err
+//  }
+//  defer session.Close()
+//
+// Executing queries
+//
+// Create queries with Session.Query. Query values should not be reused between different executions and should not be
```

Comment:
**though**: Just out of curiosity, is `should` the right word here? Shouldn't it be `must`?

Reply:
Changed to must. Well, we have Query.Bind that has documentation that says

```go
// Bind sets query arguments of query. This can also be used to rebind new query arguments
// to an existing query instance.
```

Rebinding could imply that it was intended for Query to be reusable.

On the other hand Session.Query docs states

```go
// Query generates a new query object for interacting with the database.
// Further details of the query may be tweaked using the resulting query
// value before the query is executed. Query is automatically prepared
// if it has not previously been executed.
```

which says the Query can only be modified before it is executed (which implies no reuse?).

@Zariel could you please clarify whether it is okay to say that queries must not be reused?

Reply:
Hmm, it is definitely not safe to reuse queries if speculative execution is enabled, as the goroutine running the speculative execution outlives the user's call to execute the query.

Reply:
So I guess `must` is the correct word and we can close this thread and proceed?

Reply:
Ok, merged. We can update it later in case it's not accurate.